cmd/tether/attach.go

```diff
@@ -140,78 +171,121 @@ func (t *attachServerSSH) start() error {
 	defer trace.End(trace.Begin("stop attach server"))
 
-	if t == nil || !t.enabled {
-		return
+	if t == nil {
+		err := fmt.Errorf("attach server is not configured")
+		log.Error(err)
+		return err
+	}
+
+	if !t.IsEnabled() {
+		err := fmt.Errorf("attach server is not enabled")
+		log.Error(err)
+		return err
 	}
 
-	t.enabled = false
-	conn := t.conn
-	t.conn = nil
+	log.Debugf("Setting enabled to false")
+	t.Disabled()
 
-	if conn != nil {
-		(*conn).Close()
+	log.Debugf("Canceling the context")
+	t.cancel()
+
+	log.Debugf("Acquiring the connection lock")
+	t.conn.Lock()
+	if t.conn.conn != nil {
+		log.Debugf("Closing the connection")
+		(*t.conn.conn).Close()
+		t.conn.conn = nil
 	}
+	t.conn.Unlock()
+	log.Debugf("Released the connection lock")
+
+	return nil
 }
 
 // run should not be called directly, but via start
 // run will establish an ssh server listening on the backchannel
 func (t *attachServerSSH) run() error {
 	defer trace.End(trace.Begin("main attach server loop"))
 
-	var serverConn *ssh.ServerConn
+	var serverConn struct {
+		sync.Mutex
+		*ssh.ServerConn
+	}
+	var established bool
+
 	var chans <-chan ssh.NewChannel
 	var reqs <-chan *ssh.Request
 	var err error
 
-	for t.enabled {
+	for t.IsEnabled() {
+		serverConn.Lock()
+		established = serverConn.ServerConn != nil
+		serverConn.Unlock()
+
 		// keep waiting for the connection to establish
-		for serverConn == nil {
-			// tests are passing their own connections so do not create connections when testing is set
-			if !t.testing {
-				// close the connection if required
-				if t.conn != nil {
-					(*t.conn).Close()
+		for !established && t.IsEnabled() {
+			log.Debugf("Trying to establish a connection")
+
+			establishFn := func() error {
+				t.conn.Lock()
+				defer t.conn.Unlock()
+
+				// tests are passing their own connections so do not create connections when testing is set
+				if !t.testing {
+					// close the connection if required
+					if t.conn.conn != nil {
+						(*t.conn.conn).Close()
+						t.conn.conn = nil
+					}
+					t.conn.conn, err = rawConnectionFromSerial()
+					if err != nil {
+						detail := fmt.Errorf("failed to create raw connection raw connection: %s", err)
+						log.Error(detail)
+						return detail
+					}
 				}
-				t.conn, err = rawConnectionFromSerial()
+
+				// wait for backchannel to establish
+				err = backchannel(t.ctx, t.conn.conn)
 				if err != nil {
-					detail := fmt.Errorf("failed to create raw connection raw connection: %s", err)
+					detail := fmt.Errorf("failed to establish backchannel: %s", err)
 					log.Error(detail)
-					continue
+					return detail
 				}
-			}
-			conn := t.conn
 
-			// wait for backchannel to establish
-			err = backchannel(context.Background(), conn)
-			if err != nil {
-				detail := fmt.Errorf("failed to establish backchannel: %s", err)
-				log.Error(detail)
-				continue
-			}
+				// create the SSH server
+				serverConn.Lock()
+				defer serverConn.Unlock()
```

Comment:
Ok, this _isn't_ in an anon func :-)

Here, can we Lock, defer the unlock, and come back around again and possibly deadlock ourselves?  Likewise the `defer` may never run, right?